### PR TITLE
Port all remaining library functions 

### PR DESCRIPTION
**Pull request by karoliineh**
- [x] Must be thoroughly checked by someone before merging, because the last bit was quite difficult and can likely contain mistakes
- [x] Revealed some bug: test `02/34 builtin_va_list` fails and has to be looked into, ~something to do with the autotuner~:
```
Fatal error: exception LibraryDsl.Pattern.Expected("Library function is called with more arguments than expected.")
Raised at LibraryDsl.Pattern.fail in file "src/util/library/libraryDsl.ml", line 14, characters 15-33
Called from Goblint_lib__AutoTune.hasFunction.relevant_static.(fun) in file "src/autoTune.ml", line 158, characters 41-60
Called from Stdlib__Map.Make.exists in file "map.ml", line 329, characters 29-34
Called from Goblint_lib__AutoTune.hasFunction in file "src/autoTune.ml", line 175, characters 2-77
Called from Goblint_lib__AutoTune.activateLongjmpAnalysesWhenRequired in file "src/autoTune.ml", line 210, characters 5-26
Called from Dune__exe__Goblint.main in file "src/goblint.ml", line 59, characters 6-53
Called from Stdlib.at_exit.new_exit in file "stdlib.ml", line 560, characters 59-63
Called from Stdlib.do_at_exit in file "stdlib.ml" (inlined), line 566, characters 20-61
Called from Std_exit in file "std_exit.ml", line 18, characters 8-20
See result/index.xml
```
The latter was caused by the fact that cil has special cases for handling `va_arg` and `va_start` functions. Thus, I fixed their argument definitions according to what is written in CIL and left comments.

### CODE REVIEW

File: src/util/library/libraryFunctions.ml [hunk 1]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("__open_2", unknown [drop "file" [r]; drop "oflag" []]);
+    ("__open_too_many_args", unknown []);
+    (* bzlib *)
+    ("BZ2_bzBuffToBuffCompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "blockSize100k" []; drop "verbosity" []; drop "workFactor" []]);
+    ("BZ2_bzBuffToBuffDecompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "small" []; drop "verbosity" []]);
+    (* zlib (Zebedee) *)
+    ("uncompress", unknown [drop "dest" [w]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []]);
+    ("compress2", unknown [drop "dest" [r]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []; drop "level" []]);
+    (* opensssl blowfish *)
+    ("BF_cfb64_encrypt", unknown [drop "in" [r]; drop "out" [w]; drop "length" []; drop "schedule" [r]; drop "ivec" [w]; drop "num" [w]; drop "enc" [r]]);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1204]:
```suggestion
    ("BF_cfb64_encrypt", unknown [drop "in" [r]; drop "out" [w]; drop "length" []; drop "schedule" [r]; drop "ivec" [r; w]; drop "num" [r; w]; drop "enc" []]);
```

File: src/util/library/libraryFunctions.ml [hunk 2]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    (* TODO: was "zil_replay", writes [1;2;3;5]; but couldn't find anything with 5/6 arguments, also not in bench *)
+    ("zil_replay", unknown [drop "os" [w]; drop "arg" [w]; drop "replay_func" [r]]);
+    ("_IO_getc", unknown [drop "f" [r_deep; w_deep]]);
+    ("__open_alias", unknown (drop "path" [r] :: drop "oflag" [] :: VarArgs (drop' [r])));
+    ("__open_2", unknown [drop "file" [r]; drop "oflag" []]);
+    ("__open_too_many_args", unknown []);
+    (* bzlib *)
+    ("BZ2_bzBuffToBuffCompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "blockSize100k" []; drop "verbosity" []; drop "workFactor" []]);
+    ("BZ2_bzBuffToBuffDecompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "small" []; drop "verbosity" []]);
+    (* zlib (Zebedee) *)
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1200]: These should go into zlib list then.

File: src/util/library/libraryFunctions.ml [hunk 3]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("uncompress", unknown [drop "dest" [w]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []]);
+    ("compress2", unknown [drop "dest" [r]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []; drop "level" []]);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1202]:
```suggestion
    ("compress2", unknown [drop "dest" [w]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []; drop "level" []]);
```

File: src/util/library/libraryFunctions.ml [hunk 4]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+let legacy_libs_misc_list: (string * LibraryDesc.t) list = LibraryDsl.[
+    (* TODO: was "zil_replay", writes [1;2;3;5]; but couldn't find anything with 5/6 arguments, also not in bench *)
+    ("zil_replay", unknown [drop "os" [w]; drop "arg" [w]; drop "replay_func" [r]]);
+    ("_IO_getc", unknown [drop "f" [r_deep; w_deep]]);
+    ("__open_alias", unknown (drop "path" [r] :: drop "oflag" [] :: VarArgs (drop' [r])));
+    ("__open_2", unknown [drop "file" [r]; drop "oflag" []]);
+    ("__open_too_many_args", unknown []);
+    (* bzlib *)
+    ("BZ2_bzBuffToBuffCompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "blockSize100k" []; drop "verbosity" []; drop "workFactor" []]);
+    ("BZ2_bzBuffToBuffDecompress", unknown [drop "dest" []; drop "destLen" []; drop "source" [w]; drop "sourceLen" []; drop "small" []; drop "verbosity" []]);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1199]:
```suggestion
    ("BZ2_bzBuffToBuffCompress", unknown [drop "dest" [w]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []; drop "blockSize100k" []; drop "verbosity" []; drop "workFactor" []]);
    ("BZ2_bzBuffToBuffDecompress", unknown [drop "dest" [w]; drop "destLen" [r; w]; drop "source" [r]; drop "sourceLen" []; drop "small" []; drop "verbosity" []]);
```

File: src/util/library/libraryFunctions.ml [hunk 5]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
   ]
 [@@coverage off]
 
+let legacy_libs_misc_list: (string * LibraryDesc.t) list = LibraryDsl.[
+    (* TODO: was "zil_replay", writes [1;2;3;5]; but couldn't find anything with 5/6 arguments, also not in bench *)
+    ("zil_replay", unknown [drop "os" [w]; drop "arg" [w]; drop "replay_func" [r]]);
+    ("_IO_getc", unknown [drop "f" [r_deep; w_deep]]);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1193]: Should be in Linux userspace.

File: src/util/library/libraryFunctions.ml [hunk 6]
```diff
@@ -1127,6 +1187,38 @@ let liblzma_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
   ]
 [@@coverage off]
 
+let legacy_libs_misc_list: (string * LibraryDesc.t) list = LibraryDsl.[
+    (* TODO: was "zil_replay", writes [1;2;3;5]; but couldn't find anything with 5/6 arguments, also not in bench *)
+    ("zil_replay", unknown [drop "os" [w]; drop "arg" [w]; drop "replay_func" [r]]);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1192]: Let's just get rid of it then.
```suggestion
```

File: src/util/library/libraryFunctions.ml [hunk 7]
```diff
@@ -1143,6 +1235,7 @@ let libraries = Hashtbl.of_list [
     ("pcre", pcre_descs_list);
     ("zlib", zlib_descs_list);
     ("liblzma", liblzma_descs_list);
+    ("legacy", legacy_libs_misc_list);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:1238]: Needs to be added to `lib.activated` option, both as `enum` and `default`.

File: src/util/library/libraryFunctions.ml [hunk 8]
```diff
@@ -624,7 +649,9 @@ let glibc_desc_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("__readlink_chk", unknown [drop "path" [r]; drop "buf" [w]; drop "len" []; drop "buflen" []]);
     ("__readlink_alias", unknown [drop "path" [r]; drop "buf" [w]; drop "len" []]);
     ("__overflow", unknown [drop "f" [r]; drop "ch" []]);
+    ("__ctype_b_loc", unknown []);
     ("__ctype_get_mb_cur_max", unknown []);
+    ("__maskrune", unknown [drop "c" [w]; drop "f" []]);
```
[the remark this answers is not in the document]
Comment (karoliineh) [reply src/util/library/libraryFunctions.ml:654]: Everything but this from the suggestions should be addressed now.

File: src/util/library/libraryFunctions.ml [hunk 9]
```diff
@@ -657,6 +684,18 @@ let glibc_desc_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("clntudp_create", unknown [drop "addr" [r]; drop "prognum" []; drop "versnum" []; drop "wait" [r]; drop "sockp" [w]]);
+    ("clntudp_bufcreate", unknown [drop "addr" [r]; drop "prognum" []; drop "versnum" []; drop "wait" [r]; drop "sockp" [w]; drop "sendsize" []; drop "recosize" []]);
+    ("svctcp_create", unknown [drop "sock" []; drop "send_buf_size" []; drop "recv_buf_size" []]);
+    ("authunix_create_default", unknown []);
+    ("clnt_broadcast", unknown [drop "prognum" []; drop "versnum" []; drop "procnum" []; drop "inproc" [r; c]; drop "in" [w]; drop "outproc" [r; c]; drop "out" [w]; drop "eachresult" [c]]);
+    ("clnt_sperrno", unknown [drop "stat" []]);
+    ("pmap_unset", unknown [drop "prognum" []; drop "versnum" []]);
+    ("svcudp_create", unknown [drop "sock" []]);
+    ("svc_register", unknown [drop "xprt" [r_deep; w_deep]; drop "prognum" []; drop "versnum" []; drop "dispatch" [r; w; c]; drop "protocol" []]);
+    ("svc_run", special [] Abort);
```
Comment (sim642) [inline src/util/library/libraryFunctions.ml:697]: We should leave this as `unknown` for now and maybe in a separate PR make a new kind of special (e.g. `NoReturn`) for this. If it never returns, then the following node will be dead (like `Abort`), but the program doesn't exit (so it shouldn't be `Abort`). This makes a difference for the memLeak analysis.

File: src/util/library/libraryFunctions.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sim642) [reply src/util/library/libraryFunctions.ml:654]:
```suggestion
    ("__maskrune", unknown [drop "c" []; drop "f" []]);
```

**Review by sim642 — APPROVED** [review]

**Pull request merged**